include/ghex/transport_layer/ucx/context.hpp

```diff
@@ -171,73 +178,74 @@ namespace gridtools {
+                    : m_context{c}
+                    , m_thread_primitives{tp}
+                    , m_token_ptr{t}
+                    , m_rank(c->rank())
+                    , m_size(c->size())
+                    {
+                        ucp_worker_params_t params;
+                        params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+                        params.thread_mode = mode;
+                        GHEX_CHECK_UCX_RESULT(
+                            ucp_worker_create (c->get(), &params, &m_worker.get())
+                        );
+                        ucp_address_t* worker_address;
+                        std::size_t address_length;
+                        GHEX_CHECK_UCX_RESULT(
+                            ucp_worker_get_address(m_worker.get(), &worker_address, &address_length)
+                        );
+                        m_address = address_t{
+                            reinterpret_cast<unsigned char*>(worker_address),
+                            reinterpret_cast<unsigned char*>(worker_address) + address_length};
+                        ucp_worker_release_address(m_worker.get(), worker_address);
+                        m_worker.m_moved = false;
+                    }
+
                 template<typename ThreadPrimitives>
                 const endpoint_t& worker_t<ThreadPrimitives>::connect(rank_type rank)
                 {
                     auto it = m_endpoint_cache.find(rank);
                     if (it != m_endpoint_cache.end())
                         return it->second;
-                    if (auto addr_ptr = m_context->m_db.find(rank))
-                    {
-                        auto p = m_endpoint_cache.insert(std::make_pair(rank, endpoint_t{rank, m_worker.get(), *addr_ptr}));
-                        return p.first->second;
-                    }
-                    else
-                        throw std::runtime_error("could not connect to endpoint");
+                    auto addr = m_context->m_db.find(rank);
+                    auto p = m_endpoint_cache.insert(std::make_pair(rank, endpoint_t{rank, m_worker.get(), addr}));
+                    return p.first->second;
                 }
 
             } // namespace ucx
-            
+
             template<class ThreadPrimitives>
             struct context_factory<ucx_tag, ThreadPrimitives>
             {
-                static std::unique_ptr<context<ucx_tag, ThreadPrimitives>> create(int num_threads, MPI_Comm mpi_comm)
+                template<typename DB>
+                static std::unique_ptr<context<ucx_tag, ThreadPrimitives>> create(int num_threads, DB&& db)
                 {
-                    return std::make_unique<context<ucx_tag,ThreadPrimitives>>(num_threads, mpi_comm, ucx::address_db_mpi{mpi_comm});
+                    return std::make_unique<context<ucx_tag,ThreadPrimitives>>(num_threads, db.m_mpi_comm, std::forward<DB>(db));
                 }
```

Comment:
since all the databases take a mpi_comm now, couldn't you construct the dbs inside this create function (and have the ifdefs there)?